File: archives/python/9012.py
import sys
from collections import deque
#input = sys.stdin.readline

T = int(input())
for _ in range(T):
	input_string = input()
	stack = deque([])
	for word in input_string:
		if word == "(":
			stack.append(word)
		elif word == ")":
			if len(stack) == 0 : #empty stack
				stack.append(word)
			else:
				before = stack.pop()
				if before == "(":
					continue
				else:
					stack.append(before)
					stack.append(word)
	if len(stack) == 0:
		print("YES")
	else:
		print("NO")

File: archives/cpp/9012.cpp
#include <bits/stdc++.h>
#define FOR(i, start, end) for (int i = start; i < end; i++)
#define endl "\n"
#define INF 9876543201;
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);
    
    int T; cin >> T;
    while (T--) {
        string words; cin >> words;
        stack<char> s;
        if (words.length() % 2 != 0) {
            cout << "NO" << endl;
            continue;
        }

        for (auto word: words) {
            if (word == '(') {
                s.push(word);
            } else if (word == ')') {
                if (s.empty()) {
                    s.push(word);
                } else {
                    if (s.top() == '(') {
                        s.pop();
                    } else {
                        s.push(word);
                    }
                }
            }
        }

        if (s.empty()) {
            cout << "YES" << endl;
        } else {
            cout << "NO" << endl;
        }
    }

    return 0;
}